### ISSUE TITLE
Don't know how to inference

### ISSUE THREAD

**Issue by tienduchoang** [open-edge-platform/geti#58]
hello @snosov1 @alexey-trushkov @alexey-sidnev AlexanderDokuchaev and everyone, Thank for your code. I have already trained and exported to frozen model (.pb) without optimize. Then i use frozen model for tensorrt inference server. but I got the wrong result compared to using infer.py file (checkpoints)

I don't know if my export .pb file is incorrect or my clients code are wrong

```Python
#### MY CLIENTS CODE #####
import argparse
import os
import random
import cv2
import numpy as np
import tensorflow as tf
import tensorflow.contrib.slim as slim
from builtins import range
from tensorrtserver.api import *
import tensorrtserver.api.model_config_pb2 as model_config


if __name__ == '__main__':
    input_name = 'input'
    output_name = 'd_predictions'
    model_name = 'lp-recognitor'
    protocol = ProtocolType.from_str('http')
    ctx = InferContext('localhost:8000', protocol, model_name, -1, False)

    img = cv2.imread("/data/tmp/plate/000111.png")
    img = np.float32(img)
    img = cv2.resize(img, (24, 94))

    in_frame = img.reshape((24, 94, 3)) 

    
    input_data = []
    input_data.append(in_frame) 

    results = [] 
    results.append(ctx.run(
            { input_name : input_data },
            { output_name : (InferContext.ResultFormat.RAW) }))
    print("****************results*********************", results)
```

**Comment by snosov1** [comment]
Hey, @tienduchoang !

Actually, I don't think we've ever tried tensorrt server on our side. From your description it looks like you have to report the bug to tensorrt team, since "their" results are different from local TF ("Then i use frozen model for tensorrt inference server. but I got the wrong result compared to using infer.py file")

Fair enough?

**Comment by tienduchoang** [comment]
hi @snosov1  thank for your reply. But can you continue help me.
i really need your help.

after train, i export model by using your export.py.
################ EXPORT ############################################
from __future__ import print_function
import argparse
import os
import subprocess


import tensorflow as tf
import tensorflow.contrib.slim as slim
from tensorflow.python.framework import graph_io

from lpr.trainer import inference
from utils.helpers import load_module


def parse_args():
  parser = argparse.ArgumentParser(description='Export model in IE format')
  parser.add_argument('path_to_config', help='Path to a config.py')
  # parser.add_argument('mo', help="Path to model optimizer 'mo.py' script")
  return parser.parse_args()

def export(config, tfmo, batch_size=1, precision='FP32'):
  shape = (None,) + tuple(config.input_shape) # NHWC, dynamic batch
  graph = tf.Graph()
  with graph.as_default():
    with slim.arg_scope([slim.batch_norm, slim.dropout], is_training=False):
      input_tensor = tf.placeholder(dtype=tf.float32, shape=shape, name='input')
      prob = inference(config.rnn_cells_num, input_tensor, config.num_classes)
      prob = tf.transpose(prob, (1, 0, 2))
      data_length = tf.fill([tf.shape(prob)[1]], tf.shape(prob)[0])
      result = tf.nn.ctc_greedy_decoder(prob, data_length, merge_repeated=True)
      predictions = tf.to_int32(result[0][0])
      tf.sparse_to_dense(predictions.indices, [tf.shape(input_tensor, out_type=tf.int64)[0], config.max_lp_length],
                         predictions.values, default_value=-1, name='d_predictions')
      init = tf.initialize_all_variables()
      saver = tf.train.Saver(write_version=tf.train.SaverDef.V2)

  sess = tf.Session(graph=graph)
  sess.run(init)
  checkpoints_dir = config.model_dir
  latest_checkpoint = tf.train.latest_checkpoint(checkpoints_dir)
  saver.restore(sess, latest_checkpoint)
  frozen = tf.graph_util.convert_variables_to_constants(sess, sess.graph_def, ["d_predictions"])
  tf.train.write_graph(sess.graph, os.path.join(config.model_dir, 'ie_model/'), 'graph.pbtxt', as_text=True)
  path_to_frozen_model = graph_io.write_graph(frozen, os.path.join(config.model_dir, 'ie_model/'),
                                              'graph.pb.frozen', as_text=False)


def main(_):
  args = parse_args()
  cfg = load_module(args.path_to_config)
  export(cfg, 1, 'FP32')


if __name__ == '__main__':
  tf.logging.set_verbosity(tf.logging.INFO)
  tf.app.run(main)
############################################################




Then i use .pb file to infer.
My code is: 
#################### INFER ##############################
import tensorflow as tf
import cv2


def load_graph(frozen_graph_filename):
    with tf.gfile.GFile(frozen_graph_filename, 'rb') as f:
        graph_def = tf.GraphDef()
        graph_def.ParseFromString(f.read())
    with tf.Graph().as_default() as graph:
        tf.import_graph_def(graph_def)
    return graph


if __name__ == "__main__":
    graph = load_graph(
        "/data/Videos/openvino_training_extensions/tensorflow_toolkit/lpr/chinese_lp/model/ie_model/graph.pb.frozen")
    for op in graph.get_operations():
        print(op.name)
    # output node name : import/d_predictions
    # input node name: import/input
    image = cv2.imread(
        "/data/Videos/openvino_training_extensions/data/synthetic_chinese_license_plates/crops/079437.png")
    img = cv2.resize(image, (94, 24))

    input = graph.get_tensor_by_name("import/input:0")
    output = graph.get_tensor_by_name("import/d_predictions:0")
    # print(input.shape) # [24, 94, 3]
    # print(output.shape) # [20]
    with tf.Session(graph=graph) as sess:
        results = sess.run(output, feed_dict={input: [img]})
    print(results)
#########################################################


It return [[-1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1]]
difference compared to ckpt.

**Comment by snosov1** [comment]
Hey, @tienduchoang !

Sorry, it's impossible to understand what your code is doing (one thing is that you completely strip the indentation).

**Comment by tienduchoang** [comment]
hey @snosov1  
i just export model to frozen graph (.pb file) then using .pb file to inference. (not use ckpt)

**Comment by tienduchoang** [comment]
hi sergei nosov, i am so sorry because *i bother u.*
But i really need you help.
sorry for not making you understand

After trainning, i export ckpt to frozen model (.pb). and use this attached
files to inference.
I got the result: [[-1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1
-1 -1]] but the result is different from using the ckpt model ([[11 55 61
50 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1]]).
could you explain to me?


Vào Th 3, 2 thg 4, 2019 vào lúc 17:36 Sergei Nosov <
notifications@github.com> đã viết:

**Comment by snosov1** [comment]
I've heard from @AlexanderDokuchaev that he sees the same (strange) behavior. We'll take a closer look and let you know when we have something.

**Comment by tienduchoang** [comment]
thank you. @snosov1

**Comment by AlexanderDokuchaev** [comment]
Hi @tienduchoang,
You should add preprocessing for input data in your code
```Py
    img = cv2.resize(image, (94, 24))
    img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
    img = np.float32(img)
    img = np.multiply(img, 1.0/255.0)
```
For IE model it done by [arguments for mo.py](https://github.com/opencv/openvino_training_extensions/blob/develop/tensorflow_toolkit/lpr/export.py#L38).

**Comment by tienduchoang** [comment]
ohhhh. it works. thank you very much

**Comment by tienduchoang** [comment]
hi @snosov1  @AlexanderDokuchaev , it works. thank you very much
